wrangler-core/src/main/java/io/cdap/wrangler/utils/DirectiveOutputSchemaGenerator.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.wrangler.utils;
+
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.wrangler.api.Directive;
+import io.cdap.wrangler.api.Pair;
+import io.cdap.wrangler.api.Row;
+
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * This class can be used to generate the output schema for the output data of a directive. It maintains a map of
+ * output fields present across all output rows after applying a directive. This map is used to generate the schema
+ * if the directive does not return a custom output schema.
+ */
+public class DirectiveOutputSchemaGenerator {
+  private final SchemaConverter schemaGenerator;
+  private final Map<String, Object> outputFieldMap;
+  private final Directive directive;
+
+  public DirectiveOutputSchemaGenerator(Directive directive, SchemaConverter schemaGenerator) {
+    this.directive = directive;
+    this.schemaGenerator = schemaGenerator;
+    outputFieldMap = new LinkedHashMap<>();
+  }
+
+  /**
+   * Method to add new fields from the given output to the map of fieldName --> value maintained for schema generation.
+   * A value is added to the map only if it is absent (or) if the existing value is null and given value is non-null
+   * @param output list of output {@link Row}s after applying directive.
+   */
+  public void addNewOutputFields(List<Row> output) {
+    for (Row row : output) {
+      for (Pair<String, Object> field : row.getFields()) {
+        String fieldName = field.getFirst();
+        Object fieldValue = field.getSecond();
+        if (outputFieldMap.containsKey(fieldName)) {
+          // If existing value is null, override with this non-null value
+          if (fieldValue != null && outputFieldMap.get(fieldName) == null) {
+            outputFieldMap.put(fieldName, fieldValue);
+          }
+        } else {
+          outputFieldMap.putIfAbsent(fieldName, fieldValue);
```

Comment:
nit: Any specific reason to use `putIfAbsent`? You already checked that the field is not there. Also `putIfAbsent` may work weirdly for maps with null values.